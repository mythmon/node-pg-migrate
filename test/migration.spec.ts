import type { Mock } from 'vitest';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import type { DBConnection } from '../src/db';
import { getTimestamp, loadMigrationFiles, Migration } from '../src/migration';
import type { Logger, RunnerOption } from '../src/types';

const callbackMigration = '1414549381268_names.js';
const promiseMigration = '1414549381268_names_promise.js';
const migrationsTable = 'pgmigrations';

const actionsCallback = require(`./${callbackMigration}`);
const actionsPromise = require(`./${promiseMigration}`);

describe('migration', () => {
  const dbMock = {} as DBConnection;

  const logger: Logger = {
    info: () => null,
    warn: () => null,
    error: () => null,
  };

  const options = { migrationsTable, dir: '' } as RunnerOption;

  let queryMock: Mock;

  beforeEach(() => {
    queryMock = vi.fn();
    dbMock.query = queryMock;
  });

  describe('getTimestamp', () => {
    it('should get timestamp for normal timestamp', () => {
      const now = Date.now();
      const timestamp = String(now);
      expect(getTimestamp(logger, timestamp)).toBe(now);
    });

    it('should get timestamp for shortened iso format', () => {
      const now = new Date();
      const timestamp = now.toISOString().replace(/\D/g, '');
      expect(getTimestamp(logger, timestamp)).toBe(now.valueOf());
    });

    it('should get timestamp for normal timestamp with year prefix', () => {
      const now = Date.now();
      const timestamp = `${new Date(now).getFullYear()}/${now}`;
      expect(getTimestamp(logger, timestamp)).toBe(now);
    });

    it('should get timestamp for shortened iso format with year prefix', () => {
      const now = new Date();
      const timestamp = `${new Date(now).getFullYear()}/${now.toISOString().replace(/\D/g, '')}`;
      expect(getTimestamp(logger, timestamp)).toBe(now.valueOf());
    });
  });

  describe('self.applyUp', () => {
    it('should call db.query on normal operations', async () => {
      const migration = new Migration(
        dbMock,
        callbackMigration,
        actionsCallback,
        options,
        {},
        logger
      );

      await migration.apply('up');

      expect(queryMock).toHaveBeenCalled();
    });

    it('should call db.query when returning promise on normal operations', async () => {
      const migration = new Migration(
        dbMock,
        promiseMigration,
        actionsPromise,
        options,
        {},
        logger
      );

      await migration.apply('up');

      expect(queryMock).toHaveBeenCalled();
    });

    it('should not call db.query on --dry-run', async () => {
      const migration = new Migration(
        dbMock,
        callbackMigration,
        actionsCallback,
        { ...options, dryRun: true },
        {},
        logger
      );

      await migration.apply('up');

      expect(queryMock).not.toHaveBeenCalled();
    });

    it('should make proper SQL calls', async () => {
      const migration = new Migration(
        dbMock,
        promiseMigration,
        actionsCallback,
        options,
        {},
        logger
      );

      await migration.apply('up');

      expect(queryMock).toHaveBeenCalledTimes(4);
      expect(queryMock).toHaveBeenNthCalledWith(1, 'BEGIN;');
      expect(queryMock).toHaveBeenNthCalledWith(
        2,
        expect.stringMatching('CREATE TABLE')
      );
      expect(queryMock).toHaveBeenNthCalledWith(
        3,
        expect.stringMatching(`INSERT INTO "public"."${migrationsTable}"`)
      );
      expect(queryMock).toHaveBeenNthCalledWith(4, 'COMMIT;');
    });

    it('should fail with an error message if the migration is invalid', () => {
      const invalidMigrationName = 'invalid-migration';

      const migration = new Migration(
        dbMock,
        invalidMigrationName,
        {},
        options,
        {},
        logger
      );

      const direction = 'up';

      expect(() => migration.apply(direction)).toThrow(
        new Error(
          `Unknown value for direction: ${direction}. Is the migration ${invalidMigrationName} exporting a '${direction}' function?`
        )
      );
    });
  });

  describe('self.applyDown', () => {
    it('should call db.query on normal operations', async () => {
      const migration = new Migration(
        dbMock,
        callbackMigration,
        actionsCallback,
        options,
        {},
        logger
      );

      await migration.apply('down');

      expect(queryMock).toHaveBeenCalled();
    });

    it('should not call db.query on --dry-run', async () => {
      const migration = new Migration(
        dbMock,
        callbackMigration,
        actionsCallback,
        { ...options, dryRun: true },
        {},
        logger
      );

      await migration.apply('down');

      expect(queryMock).not.toHaveBeenCalled();
    });

    it('should make proper SQL calls', async () => {
      const migration = new Migration(
        dbMock,
        promiseMigration,
        actionsCallback,
        options,
        {},
        logger
      );

      await migration.apply('down');

      expect(queryMock).toHaveBeenCalledTimes(4);
      expect(queryMock).toHaveBeenNthCalledWith(1, 'BEGIN;');
      expect(queryMock).toHaveBeenNthCalledWith(
        2,
        expect.stringMatching('DROP TABLE')
      );
      expect(queryMock).toHaveBeenNthCalledWith(
        3,
        expect.stringMatching(`DELETE FROM "public"."${migrationsTable}"`)
      );
      expect(queryMock).toHaveBeenNthCalledWith(4, 'COMMIT;');
    });
  });

  describe('loadMigrationFiles', () => {
    it('should load cockroach migration files', async () => {
      const migrations = await loadMigrationFiles('test/cockroach');
      expect(migrations).toHaveLength(12);
      expect(migrations).toMatchSnapshot();
    });

    it('should load normal migration files', async () => {
      const migrations = await loadMigrationFiles('test/migrations');
      expect(migrations).toHaveLength(90);
      expect(migrations).toMatchSnapshot();
    });

    it('should load migration files from subdirectories', async () => {
      const migrations = await loadMigrationFiles('test/migrations-subdir');
      expect(migrations).toHaveLength(4);
      expect(migrations).toMatchSnapshot();
    });
  });
});
