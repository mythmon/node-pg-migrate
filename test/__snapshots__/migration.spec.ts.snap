// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migration > loadMigrationFiles > should load cockroach migration files 1`] = `
[
  "004_table.js",
  "006_table_rename.js",
  "008_column_drop.js",
  "009_column.js",
  "011_column_rename.js",
  "012_column_alter.js",
  "014_add_constraint.js",
  "017_drop_constraint.js",
  "019_add_index.js",
  "020_create_index.js",
  "062_view.js",
  "062_view_test.js",
]
`;

exports[`migration > loadMigrationFiles > should load migration files from subdirectories 1`] = `
[
  "001/004_table.js",
  "001/006_table_rename.js",
  "002/008_column_drop.js",
  "002/009_column.js",
]
`;

exports[`migration > loadMigrationFiles > should load normal migration files 1`] = `
[
  "001_noop.js",
  "002_callback.js",
  "003_promise.js",
  "004_table.js",
  "005_table_test.js",
  "006_table_rename.js",
  "007_table_rename_test.js",
  "008_column_drop.js",
  "009_column.js",
  "010_column_test.js",
  "011_column_rename.js",
  "012_column_alter.js",
  "013_column_alter_test.js",
  "014_add_constraint.js",
  "015_add_constraint_test.js",
  "016_rename_constraint.js",
  "017_drop_constraint.js",
  "018_drop_constraint_test.js",
  "019_add_index.js",
  "020_drop_index.js",
  "021_add_type.js",
  "022_add_type_test.js",
  "023_add_type_attribute.js",
  "024_add_type_attribute_test.js",
  "025_set_type_attribute.js",
  "026_set_type_attribute_test.js",
  "027_add_type_value.js",
  "028_add_type_value_test.js",
  "029_rename_type_attribute.js",
  "030_rename_type_attribute_test.js",
  "031_drop_type_attribute.js",
  "032_drop_type_attribute_test.js",
  "033_drop_type.js",
  "034_drop_type_test.js",
  "035_role_add.js",
  "036_role_alter.js",
  "037_role_rename.js",
  "038_role_drop.js",
  "039_function_create.js",
  "040_function_rename.js",
  "041_function_test.js",
  "042_function_drop.js",
  "043_trigger_create_rename.js",
  "044_trigger_test.js",
  "045_trigger_drop.js",
  "046_domain_create_rename.js",
  "047_domain_check.js",
  "048_domain_drop.js",
  "049_sequence_create_rename.js",
  "050_sequence_test.js",
  "051_sequence_alter.js",
  "052_sequence_alter_test.js",
  "053_sequence_drop.js",
  "054_operator_create.js",
  "055_operator_test.js",
  "056_operator_drop.js",
  "057_policy_create.js",
  "058_policy_test.js",
  "059_policy_drop.js",
  "060_column_comment.js",
  "061_column_comment_test.js",
  "062_view.js",
  "063_view_test.js",
  "064_alter_view_column.js",
  "065_materialized_view.js",
  "066_materialized_view_test.js",
  "067_materialized_view_alter.js",
  "068_extension.js",
  "069_comments.js",
  "070_extension_schema.js",
  "071_constraint_name_for_foreign_key.js",
  "072_alter_column_comment.js",
  "073_alter_column_comment_test.js",
  "074_rename_type_value.js",
  "075_drop_index_schema.js",
  "076_create_table_like.js",
  "077_create_table_generated_column.js",
  "078_add_column_if_not_exists.js",
  "079_drop_index_schema.js",
  "080_create_table_generated_column_take_2.js",
  "081_temporary_table.js",
  "082_view_options.js",
  "083_alter_column_sequence.js",
  "084_drop_unique_index.js",
  "085_grant_tables_schemas_roles.js",
  "086_grant_test.js",
  "087_revoke_tables_schemas_roles.js",
  "088_revoke_test.js",
  "089_grant_reverse.js",
  "090_create_cast.js",
]
`;
